toplevel/coqinit.ml

```diff
@@ -104,6 +104,9 @@ let libs_init_load_path ~load_init =
   [build_stdlib_path ~load_init ~unix_path:(coqlib/"theories") ~coq_path ~with_ml:false;
    build_stdlib_path ~load_init ~unix_path:(coqlib/"plugins")  ~coq_path ~with_ml:true ] @
 
+  (* then vendored plugins *)
+  [build_userlib_path ~unix_path:(coqlib/"vendor")] @
```

Comment:
If possible I'd like to avoid more hardcoded paths.

Reply:
That's what I tried initially, but I'm not sure I see the full story: I want Ltac 1 and 2 usable without passing flags to `coqtop` or `coqide`. How do I achieve that without hardcoded paths?

Of course, I wouldn't mind a future change where we remove all hardcoded paths (and leave it to interfaces to pass the right paths for example), but this should be done at the same time for Ltac 1 and 2 (and other plugins).

Reply:
> That's what I tried initially, but I'm not sure I see the full story: I want Ltac 1 and 2 usable without passing flags to `coqtop` or `coqide`. How do I achieve that without hardcoded paths?

Oh, if that's what you want I dunno, then it should maybe be merged into the stdlib? But I am not sure it is a good idea; IMHO you should add `ltac2` to your `_CoqProject` or `dune` file and let coqide / PG use the correct paths.

Implicit dependencies are really bad as for example they break build compositionality; I had to add special code to detect when Coq's stdlib is in the local build scope as indeed, due to the implicit dependency on `stdlib` things would work otherwise when composing with Coq.

I mean, if you have your library `foo` that depends on the implicit stdlib, you cannot generate dep rules for `stdlib` indeed, when it is implicitly loaded, but now you compose with it and you have to see it as an explicit dep.

Reply:
That's still to discuss, but for the Dune 2.0 version of the Coq language we may indeed require to add `stdlib` explicitly to the list of dependencies.

Reply:
Anyways I won't oppose to this, if we can avoid the better, but if in the end people do prefer to have ltac2 implicit that's OK I think. But that brings into question scalability of the `vendor` approach and even namespace pollution.

Reply:
Note that `vendor` is given a `-Q` semantics, so that limits a lot the namespace pollution compared to putting it in `plugins`. Each plugin in `vendor` lives under a separate root.

I'm all in for an approach where all dependencies are explicit, but I'd rather do it for everything at the same time (including stdlib indeed). Toolchains will need to be adapted. Meanwhile, the goal of this PR is precisely to make Ltac2 as simple to load as Ltac1, to stimulate the transition and be able to move away from Ltac1 quirks as soon as possible.

Reply:
Then if that is desired ltac2 should follow the default install procedure and go to user-contribs, as the non-vendored version.

Reply:
Namespace pollution is the same as the default setup where things go to user-contrib, but this makes it a bit worse as now you have a namespace collusion.

Reply:
> Then if that is desired ltac2 should follow the default install procedure and go to user-contribs, as the non-vendored version.

I can try that.

> this makes it a bit worse as now you have a namespace collusion.

What namespace collusion?

Reply:
You can have user-contribs/A and vendor/A

Reply:
Yeah so I guess the best for compatibility is that the vendored and non-vendored version install the same. This is also the approach that works best for dune.

Reply:
> the vendored and non-vendored version

Just to make sure we are on the same page: there won't be a "non-vendored" version of Ltac2, so I guess you mean earlier versions.

Reply:
I'm not sure what you want me to do for local `make`-based builds, though. All build outputs are produced alongside source files...

Reply:
Yeah the earlier version is the non-vendored one; but from the point of view of a library that uses ltac2 there is a non-vendored version if coq 8.10 is in scope and a vendored version if in 8.11.

Reply:
But we should indeed use `in-tree` and `out-of-tree`, that's the correct terminology here.

Reply:
> I'm not sure what you want me to do for local `make`-based builds, though. All build outputs are produced alongside source files...

I am not sure anything special has to be done? Local users of `ltac2` should pass the `-R` flag to the vendor dir, but for example in the test-suite this is already done for the stdlib even.

Reply:
So when the stdlib is not implicit we will replace `-coqlib foo` by `-Q foo Coq`.

Reply:
> I am not sure anything special has to be done? Local users of `ltac2` should pass the `-Q` flag to the vendor dir, but for example in the test-suite this is already done even or the stdlib.

I'm lost. Local users of stdlib certainly don't need to pass `-R theories Coq` to `coqtop` today. How do I achieve that for Ltac2 in your view?

Reply:
They do pass `-coqlib foo` in most cases; for example the test suite, etc... otherwise things don't work in all cases.

If you are referring to interactive use you could use the "-local" flag, but IMHO I dread that flag. Having a non-uniform layout is a real pain.

You could use a wrapper I guess; in the Dune case things will work out of the box once `user-contrib` is in scope.

Reply:
> Local users of stdlib certainly don't need to pass `-R theories Coq` to `coqtop` today.

Note that if you build in non-local mode and depending on your OS then you have to either install Coq or use `-coqlib`. So the current design is not something robust enough as to use it as a reference.

Reply:
I think we are mixing two topics here. Having a uniform design for building in-tree and out-of-tree dependencies is desirable, but it's not there yet. For the scope of this PR, I'd like to move Ltac2 in-tree, with all the convenience that Ltac1 has *in the current design*. That means loading `coqtop` built with `-local` from PG without any problem finding Ltac2, and so on.

I will strongly support cleaning this design and removing all hard coded paths, but as a second step.

Remember that the only reason we are not putting Ltac2 in the `Coq` namespace is that stdlib is loaded with `-R`, not `-Q`, so even clashes with short names are problematic.

Reply:
> Remember that the only reason we are not putting Ltac2 in the `Coq` namespace is that stdlib is loaded with `-R`, not `-Q`, so even clashes with short names are problematic.

In fact, I'm starting to think I'd have a much easier time renaming `String.v` to `Ltac2String.v` etc, and dropping everything in stdlib.

Reply:
> That means loading `coqtop` built with `-local` from PG without any problem finding Ltac2, and so on.

That case is already poorly supported, but you can just add a `_CoqProject` file as is done for the stdlib?

I tho insist that `-local` is a very special case and should not drive the design here; what most users see is not local, and that case has to be right first.

> Remember that the only reason we are not putting Ltac2 in the `Coq` namespace is that stdlib is loaded with `-R`, not `-Q`, so even clashes with short names are problematic.

Well, you have this problem in both approaches, don't you?

> In fact, I'm starting to think I'd have a much easier time renaming `String.v` to `Ltac2String.v` etc, and dropping everything in stdlib.

Yeah you could move `Ltac2` under Coq, but experience with `math-comp` tells me that was a very bad idea, so I'd rather not do that. It seems like using a bridge-sized sledgehammer to crack a nut.

The current design where `ltac2` is in `vendor` and conserves the prefix is OK; dropping the added path is very easy, just guard it under `Coq_config.local` and then non-local builds will behave in a backwards compatible-way.

Reply:
I'll try to summarize the discussion in the review as this went for too long.

Reply:
A last note regarding `-R` vs `-Q`: for quite a lot of time I used `-Q` as the default for the standard library in SerAPI jsCoq; it caused some pain to users so in the end I did revert it back.

However maybe the time is now to make `-Q` the default, and emit a warning in 8.11 for anyone using an unqualified module.

Reply:
> using an unqualified module.

an exception to the warning should usually be when compiling files from the same namespace than the unqualified file.